internal/pkg/redis/xpert.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (C) 2023 IOTech Ltd
+
+package redis
+
+import (
+	"github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg"
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
+)
+
+func (c Client) PublishBinaryData(data []byte, topic string) error {
+	if c.publishClient == nil {
+		return pkg.NewMissingConfigurationErr("PublishHostInfo", "Unable to create a connection for publishing")
+	}
+
+	if topic == "" {
+		// Empty topics are not allowed for Redis
+		return pkg.NewInvalidTopicErr("", "Unable to publish to the invalid topic")
+	}
+
+	topic = convertToRedisTopicScheme(topic)
+	return c.publishClient.SendBinaryData(topic, data)
+}
+
+func (c Client) SubscribeBinaryData(topics []types.TopicChannel, messageErrors chan error) error {
+	if c.subscribeClient == nil {
+		return pkg.NewMissingConfigurationErr("SubscribeHostInfo", "Unable to create a connection for subscribing")
+	}
+
+	err := c.validateTopics(topics)
+	if err != nil {
+		return err
+	}
+
+	for i := range topics {
+		go func(topic types.TopicChannel) {
+			topicName := convertToRedisTopicScheme(topic.Topic)
+			messageChannel := topic.Messages
+			for {
+				message, err := c.subscribeClient.ReceiveBinaryData(topicName)
+				if err != nil {
+					messageErrors <- err
+					continue
+				}
+
+				message.ReceivedTopic = convertFromRedisTopicScheme(message.ReceivedTopic)
+
+				messageChannel <- *message
+			}
+		}(topics[i])
+	}
+
+	return nil
+}
+
+func (g *goRedisWrapper) SendBinaryData(topic string, data []byte) error {
+	_, err := g.wrappedClient.Publish(topic, data).Result()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *goRedisWrapper) ReceiveBinaryData(topic string) (*types.MessageEnvelope, error) {
+	subscription := g.getSubscription(topic)
+	data, err := subscription.ReceiveMessage()
+	if err != nil {
+		return nil, err
+	}
+	// Use MessageEnvelope.Payload to store the binary data instead of unmarshalling binary to MessageEnvelope
+	messageEnvelope := types.NewMessageEnvelopeForRequest([]byte(data.Payload), nil)
```

Comment:
why are you using NewMessageEnvelopeForRequest, not NewMessageEnvelope?  It's a receiving function.
again, create a context and generate corelation-id

Reply:
I thought `NewMessageEnvelopeForRequest` can provide the required field without creating a context.
```
// NewMessageEnvelope creates a new MessageEnvelope for the specified payload with attributes from the specified context
func NewMessageEnvelope(payload []byte, ctx context.Context) MessageEnvelope {
	envelope := MessageEnvelope{
		Versionable:   commonDTO.NewVersionable(),
		CorrelationID: fromContext(ctx, common.CorrelationHeader),
		ContentType:   fromContext(ctx, common.ContentType),
		Payload:       payload,
		QueryParams:   make(map[string]string),
	}

	return envelope
}

// NewMessageEnvelopeForRequest creates a new MessageEnvelope for sending request to EdgeX via internal
// MessageBus to target Device Service. Used when request is from internal App Service via command client.
func NewMessageEnvelopeForRequest(payload []byte, queryParams map[string]string) MessageEnvelope {
	envelope := MessageEnvelope{
		CorrelationID: uuid.NewString(),
		Versionable:   commonDTO.NewVersionable(),
		RequestID:     uuid.NewString(),
		ErrorCode:     0,
		Payload:       payload,
		ContentType:   common.ContentTypeJSON,
		QueryParams:   make(map[string]string),
	}

	if len(queryParams) > 0 {
		envelope.QueryParams = queryParams
	}

	return envelope
}
```